File: tools/issue_contract_repair.py
#!/usr/bin/env python3
"""Issue-contract repair-assist helper — turns a contract-gate HOLD into a
reviewable repair manifest, one row per issue.

`fak issue contract` / `tools/issue_resolve_dispatch.py` (floor
DEFAULT_ISSUE_CONTRACT_MIN_SCORE) refuse to spawn a worker on an issue whose
17-field structured contract (parent_ref, done_condition, acceptance_gate,
work_unit, expected_steps, ...) is incomplete. As of 2026-06-30 that gate holds
the ENTIRE open backlog (0/342 routed via `fak dispatch route --json`) because
the schema was introduced the same commit as the gate itself — the backlog
predates it, not because the issues are badly chosen.

This helper does NOT bypass the gate and does NOT edit GitHub. It classifies
each candidate issue's contract-review reasons into a repair KIND (mirroring
`issueContractRepairKinds` in cmd/fak/issue_contract.go), and for each kind
either:
  - computes a deterministic, zero-authorship-risk fix (`template`: the
    already-built normalized-header repair for a corrupted generation
    marker), or
  - proposes a lane suggestion from the heuristic router (`route`), or
  - lists exactly the missing fields with a one-line human question per
    field (`scope`/`noise`/`split`/`private`/`other`) -- it NEVER invents the
    field content.

Every action's `cmd` is always `None`: applying a proposal (via `gh issue
edit`) stays a manual, operator-approved step, same discipline as
`tools/issue_triage.py`'s review-only actions.

Usage:
    python tools/issue_contract_repair.py --lane docs --limit 10 --json
    python tools/issue_contract_repair.py --limit 50 --markdown --out docs/_audits/issue-contract-repairs-2026-07-01.md
    python tools/issue_contract_repair.py --actions --out docs/_audits/issue-contract-repairs-2026-07-01.json

Exit codes: 0 = ran clean (including "no candidates") · 2 = infra error.
"""
from __future__ import annotations

import argparse
import datetime as dt
import json
import subprocess
import sys
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parent))
from dispatch_worker import install_no_window_subprocess_defaults  # noqa: E402
install_no_window_subprocess_defaults(subprocess)

import issue_resolve_dispatch as ird  # noqa: E402
import issue_lane_router as ilr  # noqa: E402

FETCH_CAP = 500  # generous over the ~342-issue current backlog; not a CLI knob

# ---------------------------------------------------------------------------
# Repair-kind classification -- mirrors cmd/fak/issue_contract.go's
# issueContractRepairKinds()/issueContractRepairAction() exactly (same reason
# constants, same order-preserving dedup, same default->"other" fallback) so
# a Python-side manifest and a `fak issue contract` JSON dump agree on kind.
# ---------------------------------------------------------------------------
_REASON_KIND = {
    "ISSUE_NOT_DISPATCH_LEAF": "split",
    "ISSUE_OVERSIZED_EXPECTED_STEPS": "split",
    "ISSUE_SCOPE_INCOMPLETE": "scope",
    "ISSUE_UNROUTED": "route",
    "ISSUE_LIVE_UNARMORED": "noise",
    "ISSUE_NOISE_CONTROL_INCOMPLETE": "noise",
    "ISSUE_AGENT_CONTEXT_INCOMPLETE": "noise",
    "ISSUE_PRIVATE_BOUNDARY": "private",
    "ISSUE_UNEXPANDED_TEMPLATE": "template",
}
_KIND_RANK = {"split": 1, "scope": 2, "route": 3, "noise": 4, "private": 5,
              "template": 6, "other": 9}
_KIND_ACTION = {
    "split": "decompose each non-leaf or oversized row into child issues within the dispatch step budget",
    "scope": "add the missing parent/current-state/scope/done/witness/closure fields before dispatch",
    "route": "add a lane or path hints section so the issue maps to one dispatch lane",
    "noise": "add trigger, batch policy, agent context, and live dedupe/cap evidence before automated sync",
    "private": "remove private/operator-only evidence or move the work to the private companion repo",
    "template": "dry-run a normalized generated-header repair, review it, then apply explicitly if accepted",
    "other": "inspect the review reasons and repair the row before dispatch",
}
# One-line human question per missing field. Never invents the answer -- this
# is the question a human (or a follow-up agent under human review) answers.
_FIELD_QUESTION = {
    "parent_ref": "What epic/parent issue (if any) does this belong under?",
    "current_state": "What is the current state of the code/system before this change?",
    "why_now": "Why does this need to happen now, not later?",
    "working_spine": "What is the smallest end-to-end path (the working spine) this change moves?",
    "in_scope": "What is explicitly in scope for this issue?",
    "out_of_scope": "What is explicitly out of scope (so a worker doesn't over-reach)?",
    "done_condition": "What observable state means this issue is done?",
    "acceptance_gate": "What check/command proves the done condition is met?",
    "closure_binding": "What commit-message convention binds the closing commit to this issue (e.g. `#N` in the subject)?",
    "work_unit": "What is the single unit of work here (one leaf, one commit)?",
    "expected_steps": "Roughly how many discrete steps should this take (a size estimate)?",
    "assumptions": "What is being assumed that, if wrong, would change the approach?",
    "confusion_risks": "What could a worker misunderstand or conflate here?",
    "coordination": "Does this touch a lane/file another worker might also be editing?",
    "trigger": "What event or condition should cause this to be picked up?",
    "batch_policy": "Should this run standalone or as part of a batch with related issues?",
    "witness": "What evidence (log, diff, test) proves this was actually done?",
}


def repair_kinds(reasons: list[str]) -> list[str]:
    """Order-preserving, deduped reason->kind fold. Empty/unmapped -> ["other"]."""
    kinds: list[str] = []
    for reason in reasons:
        kind = _REASON_KIND.get(reason, "other")
        if kind not in kinds:
            kinds.append(kind)
    return kinds or ["other"]


def primary_kind(kinds: list[str]) -> str:
    return min(kinds, key=lambda k: _KIND_RANK.get(k, 9))


def repair_action(kind: str) -> str:
    return _KIND_ACTION.get(kind, _KIND_ACTION["other"])


def field_scaffold(missing_fields: list[str]) -> list[dict[str, str]]:
    return [{"field": f, "question": _FIELD_QUESTION.get(f, f"What is the missing '{f}'?")}
            for f in missing_fields]


# ---------------------------------------------------------------------------
# Data sources
# ---------------------------------------------------------------------------

def fetch_open_issues(workspace: Path, *, fetch_cap: int = FETCH_CAP) -> list[dict[str, Any]]:
    """All open issues (number/title/body/labels), oldest (lowest number) first.
    `gh issue list` returns newest-first; this reverses it on purpose -- same
    ordering decision `issue_resolve_dispatch.lane_issue_numbers` makes."""
    proc = subprocess.run(
        ["gh", "issue", "list", "--state", "open", "--limit", str(fetch_cap),
         "--json", "number,title,body,labels"],
        cwd=str(workspace), capture_output=True, text=True, encoding="utf-8",
        errors="replace", timeout=60)
    if proc.returncode != 0:
        raise RuntimeError(f"gh issue list failed (rc={proc.returncode}): "
                           f"{(proc.stderr or '').strip() or 'no stderr'}")
    issues = json.loads(proc.stdout or "[]")
    return sorted(issues, key=lambda i: int(i.get("number") or 0))


def template_repair_plan(workspace: Path, issue: dict[str, Any],
                         number: int) -> dict[str, Any] | None:
    """Targeted live call for one issue's already-computed template repair
    (`internal/issuecontract.BuildTemplateRepairPlan`, dry-run only). Returns
    the matching `template_repair_plans[]` entry, or None if the Go side found
    nothing to propose."""
    import tempfile
    tmp: Path | None = None
    try:
        with tempfile.NamedTemporaryFile("w", encoding="utf-8", suffix=".json",
                                         delete=False) as f:
            tmp = Path(f.name)
            json.dump([ird._issue_record_for_contract(issue, number)], f, ensure_ascii=False)
            f.write("\n")
        cmd = [*ird._fak_command_prefix(workspace), "issue", "contract",
               "--from-issues", str(tmp), "--live", "--json"]
        proc = subprocess.run(
            cmd, cwd=str(workspace), capture_output=True, text=True, encoding="utf-8",
            errors="replace", timeout=60, creationflags=ird.no_window_creationflags())
    except (OSError, subprocess.SubprocessError):
        return None
    finally:
        if tmp is not None:
            try:
                tmp.unlink()
            except OSError:
                pass
    try:
        doc = json.loads(proc.stdout or "{}")
    except (TypeError, ValueError):
        return None
    for plan in (doc.get("template_repair_plans") or []):
        if isinstance(plan, dict) and int(plan.get("issue_number") or plan.get("number") or -1) == number:
            return plan
    return None


def build_repair_row(workspace: Path, issue: dict[str, Any],
                     concurrent: list[str], trees: dict[str, list[str]]) -> dict[str, Any] | None:
    """One issue -> one manifest row, or None if the contract already passes
    (nothing to repair)."""
    number = int(issue.get("number") or 0)
    contract = ird.issue_contract_review(workspace, issue, number)
    score = int(contract.get("score") or 0)
    if contract.get("ok") and score >= ird.DEFAULT_ISSUE_CONTRACT_MIN_SCORE:
        return None

    review = contract.get("review") if isinstance(contract.get("review"), dict) else {}
    reasons = [str(r) for r in (review.get("reasons") or []) if r]
    missing_fields = [str(m) for m in (review.get("missing_fields") or []) if m]
    kinds = repair_kinds(reasons)
    kind = primary_kind(kinds)

    row: dict[str, Any] = {
        "number": number,
        "title": str(issue.get("title") or "")[:120],
        "score": score,
        "reasons": reasons,
        "kinds": kinds,
        "kind": kind,
        "next_action": repair_action(kind),
        "ready": False,
        "proposed_lane": None,
        "route_confidence": None,
        "proposed_header": None,
        "missing_fields": field_scaffold(missing_fields) if kind in
            ("scope", "noise", "split", "private", "other") else [],
    }

    if "route" in kinds:
        route = ilr.route_issue(issue, concurrent, trees)
        if route.get("lane"):
            row["proposed_lane"] = route["lane"]
            row["route_confidence"] = route.get("confidence")

    if "template" in kinds:
        plan = template_repair_plan(workspace, issue, number)
        header = str((plan or {}).get("proposed_normalized_header") or "").strip()
        if header:
            row["ready"] = True
            row["proposed_header"] = header

    return row


# ---------------------------------------------------------------------------
# Manifest + rendering
# ---------------------------------------------------------------------------

def build_manifest(workspace: Path, *, lane: str | None, limit: int,
                   as_of: str) -> dict[str, Any]:
    concurrent, trees = ilr.lane_taxonomy(workspace)
    issues = fetch_open_issues(workspace)

    if lane:
        filtered = []
        for issue in issues:
            route = ilr.route_issue(issue, concurrent, trees)
            if route.get("lane") == lane or route.get("blocked_lane") == lane:
                filtered.append(issue)
        issues = filtered

    issues = issues[:limit]

    rows: list[dict[str, Any]] = []
    for issue in issues:
        row = build_repair_row(workspace, issue, concurrent, trees)
        if row is not None:
            rows.append(row)

    by_kind: dict[str, int] = {}
    for r in rows:
        by_kind[r["kind"]] = by_kind.get(r["kind"], 0) + 1

    return {
        "schema": "fak.issue-contract-repair.v1",
        "as_of": as_of,
        "workspace": str(workspace),
        "lane": lane,
        "limit": limit,
        "counts": {
            "candidates_examined": len(issues),
            "needs_repair": len(rows),
            "ready": sum(1 for r in rows if r["ready"]),
            "by_kind": by_kind,
        },
        "issues": rows,
    }


def build_actions(manifest: dict[str, Any]) -> list[dict[str, Any]]:
    """Lean, review-only action list. `cmd` is ALWAYS None -- this tool never
    emits a `gh issue edit/comment/close/assign` command; applying a proposal
    stays a manual, operator-approved step."""
    return [{
        "number": r["number"],
        "kind": r["kind"],
        "ready": r["ready"],
        "reason": ", ".join(r["reasons"]) or "issue contract below spawn floor",
        "next_action": r["next_action"],
        "cmd": None,
    } for r in manifest["issues"]]


def render_markdown(manifest: dict[str, Any]) -> str:
    c = manifest["counts"]
    lane = manifest["lane"] or "(all lanes)"
    L = [
        f"# Issue-contract repairs — {manifest['as_of']}",
        "",
        f"**Lane:** {lane}  ·  **examined:** {c['candidates_examined']}  ·  "
        f"**needs repair:** {c['needs_repair']}  ·  **template-ready:** {c['ready']}",
        "",
        "> Read-only pass. Never edits, labels, comments on, or closes an issue. "
        "`template`-kind rows carry a dry-run-computed header fix; every other "
        "kind lists the missing fields as questions for a human/agent to answer "
        "-- content is never invented here.",
        "",
        "## Counts by kind",
        "",
        "| kind | count |",
        "|---|---:|",
    ]
    for kind, n in sorted(c["by_kind"].items(), key=lambda kv: _KIND_RANK.get(kv[0], 9)):
        L.append(f"| {kind} | {n} |")
    L.append("")
    L.append("## Rows")
    L.append("")
    L.append("| # | kind | ready | score | title |")
    L.append("|---|---|---|---:|---|")
    for r in manifest["issues"]:
        L.append(f"| #{r['number']} | {r['kind']} | {r['ready']} | {r['score']} | {r['title']} |")
    L.append("")
    return "\n".join(L)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Issue-contract repair-assist manifest (read-only).")
    ap.add_argument("--workspace", default=None, help="repo root (default: this tool's repo)")
    ap.add_argument("--lane", default=None, help="restrict to one dispatch lane")
    ap.add_argument("--limit", type=int, default=50,
                    help="max issues to examine, oldest issue number first (default: 50)")
    ap.add_argument("--json", action="store_true")
    ap.add_argument("--markdown", action="store_true")
    ap.add_argument("--actions", action="store_true")
    ap.add_argument("--out", default=None, help="write output to this path")
    ap.add_argument("--as-of", default=None, help="date stamp (default: today UTC)")
    a = ap.parse_args(argv)

    workspace = Path(a.workspace).resolve() if a.workspace else ird.repo_root()
    as_of = a.as_of or dt.datetime.now(dt.timezone.utc).date().isoformat()

    try:
        manifest = build_manifest(workspace, lane=a.lane, limit=a.limit, as_of=as_of)
    except (RuntimeError, OSError) as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        return 2

    if a.actions:
        rendered = json.dumps({"as_of": as_of, "actions": build_actions(manifest)}, indent=2)
    elif a.markdown:
        rendered = render_markdown(manifest)
    else:
        rendered = json.dumps(manifest, indent=2)

    if a.out:
        p = Path(a.out)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(rendered + "\n", encoding="utf-8")
        print(f"wrote {p}")
    else:
        print(rendered)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/issuecontractrepair/issuecontractrepair.go
package issuecontractrepair

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"sort"
	"strings"

	"github.com/anthony-chaudhary/fak/internal/issuecontract"
	"github.com/anthony-chaudhary/fak/internal/windowgate"
)

const (
	Schema     = "fak.issue-contract-repair.v1"
	DefaultCap = 500
	MinScore   = 100
)

var reasonKind = map[string]string{
	issuecontract.ReasonNotDispatchLeaf:    "split",
	issuecontract.ReasonOversizedSteps:     "split",
	issuecontract.ReasonScopeIncomplete:    "scope",
	issuecontract.ReasonUnrouted:           "route",
	issuecontract.ReasonLiveUnarmored:      "noise",
	issuecontract.ReasonNoiseIncomplete:    "noise",
	issuecontract.ReasonAgentIncomplete:    "noise",
	issuecontract.ReasonPrivateBoundary:    "private",
	issuecontract.ReasonUnexpandedTemplate: "template",
}

var kindRank = map[string]int{"split": 1, "scope": 2, "route": 3, "noise": 4, "private": 5, "template": 6, "other": 9}

var kindAction = map[string]string{
	"split":    "decompose each non-leaf or oversized row into child issues within the dispatch step budget",
	"scope":    "add the missing parent/current-state/scope/done/witness/closure fields before dispatch",
	"route":    "add a lane or path hints section so the issue maps to one dispatch lane",
	"noise":    "add trigger, batch policy, agent context, and live dedupe/cap evidence before automated sync",
	"private":  "remove private/operator-only evidence or move the work to the private companion repo",
	"template": "dry-run a normalized generated-header repair, review it, then apply explicitly if accepted",
	"other":    "inspect the review reasons and repair the row before dispatch",
}

var fieldQuestion = map[string]string{
	"parent_ref":      "What epic/parent issue (if any) does this belong under?",
	"current_state":   "What is the current state of the code/system before this change?",
	"why_now":         "Why does this need to happen now, not later?",
	"working_spine":   "What is the smallest end-to-end path (the working spine) this change moves?",
	"in_scope":        "What is explicitly in scope for this issue?",
	"out_of_scope":    "What is explicitly out of scope (so a worker doesn't over-reach)?",
	"done_condition":  "What observable state means this issue is done?",
	"acceptance_gate": "What check/command proves the done condition is met?",
	"closure_binding": "What commit-message convention binds the closing commit to this issue (e.g. `#N` in the subject)?",
	"work_unit":       "What is the single unit of work here (one leaf, one commit)?",
	"expected_steps":  "Roughly how many discrete steps should this take (a size estimate)?",
	"assumptions":     "What is being assumed that, if wrong, would change the approach?",
	"confusion_risks": "What could a worker misunderstand or conflate here?",
	"coordination":    "Does this touch a lane/file another worker might also be editing?",
	"trigger":         "What event or condition should cause this to be picked up?",
	"batch_policy":    "Should this run standalone or as part of a batch with related issues?",
	"witness":         "What evidence (log, diff, test) proves this was actually done?",
}

type FieldPrompt struct {
	Field    string `json:"field"`
	Question string `json:"question"`
}

type RouteProposal struct {
	Lane        string
	BlockedLane string
	Confidence  string
}

type RepairRow struct {
	Number          int           `json:"number"`
	Title           string        `json:"title"`
	Score           int           `json:"score"`
	Reasons         []string      `json:"reasons"`
	Kinds           []string      `json:"kinds"`
	Kind            string        `json:"kind"`
	NextAction      string        `json:"next_action"`
	Ready           bool          `json:"ready"`
	ProposedLane    *string       `json:"proposed_lane"`
	RouteConfidence *string       `json:"route_confidence"`
	ProposedHeader  *string       `json:"proposed_header"`
	MissingFields   []FieldPrompt `json:"missing_fields"`
}

type Counts struct {
	CandidatesExamined int            `json:"candidates_examined"`
	NeedsRepair        int            `json:"needs_repair"`
	Ready              int            `json:"ready"`
	ByKind             map[string]int `json:"by_kind"`
}

type Manifest struct {
	Schema    string      `json:"schema"`
	AsOf      string      `json:"as_of"`
	Workspace string      `json:"workspace"`
	Lane      *string     `json:"lane"`
	Limit     int         `json:"limit"`
	Counts    Counts      `json:"counts"`
	Issues    []RepairRow `json:"issues"`
}

type Action struct {
	Number     int     `json:"number"`
	Kind       string  `json:"kind"`
	Ready      bool    `json:"ready"`
	Reason     string  `json:"reason"`
	NextAction string  `json:"next_action"`
	Cmd        *string `json:"cmd"`
}

type Options struct {
	Lane     string
	Limit    int
	AsOf     string
	MinScore int
	Review   func(issuecontract.IssueDraft) issuecontract.Review
	Route    func(issuecontract.IssueDraft, issuecontract.Review) RouteProposal
	Template func(issuecontract.IssueDraft) (issuecontract.TemplateRepairPlan, bool)
}

func RepairKinds(reasons []string) []string {
	var kinds []string
	seen := map[string]bool{}
	for _, reason := range reasons {
		kind := reasonKind[reason]
		if kind == "" {
			kind = "other"
		}
		if !seen[kind] {
			seen[kind] = true
			kinds = append(kinds, kind)
		}
	}
	if len(kinds) == 0 {
		return []string{"other"}
	}
	return kinds
}

func PrimaryKind(kinds []string) string {
	if len(kinds) == 0 {
		return "other"
	}
	best := kinds[0]
	for _, kind := range kinds[1:] {
		if rank(kind) < rank(best) {
			best = kind
		}
	}
	return best
}

func RepairAction(kind string) string {
	if action := kindAction[kind]; action != "" {
		return action
	}
	return kindAction["other"]
}

func FieldScaffold(missing []string) []FieldPrompt {
	out := make([]FieldPrompt, 0, len(missing))
	for _, field := range missing {
		question := fieldQuestion[field]
		if question == "" {
			question = fmt.Sprintf("What is the missing '%s'?", field)
		}
		out = append(out, FieldPrompt{Field: field, Question: question})
	}
	return out
}

func BuildRepairRow(issue issuecontract.IssueDraft, review issuecontract.Review, route RouteProposal, template issuecontract.TemplateRepairPlan, hasTemplate bool, minScore int) *RepairRow {
	if minScore <= 0 {
		minScore = MinScore
	}
	if review.OK && review.Score.Total >= minScore {
		return nil
	}
	reasons := append([]string(nil), review.Reasons...)
	kinds := RepairKinds(reasons)
	kind := PrimaryKind(kinds)
	row := &RepairRow{
		Number:     issue.Number,
		Title:      truncate(strings.TrimSpace(issue.Title), 120),
		Score:      review.Score.Total,
		Reasons:    reasons,
		Kinds:      kinds,
		Kind:       kind,
		NextAction: RepairAction(kind),
		Ready:      false,
	}
	if scaffoldsMissingFields(kind) {
		row.MissingFields = FieldScaffold(review.MissingFields)
	} else {
		row.MissingFields = []FieldPrompt{}
	}
	if contains(kinds, "route") && strings.TrimSpace(route.Lane) != "" {
		lane := strings.TrimSpace(route.Lane)
		row.ProposedLane = &lane
		if strings.TrimSpace(route.Confidence) != "" {
			conf := strings.TrimSpace(route.Confidence)
			row.RouteConfidence = &conf
		}
	}
	if contains(kinds, "template") && hasTemplate && strings.TrimSpace(template.ProposedNormalizedHeader) != "" {
		header := strings.TrimSpace(template.ProposedNormalizedHeader)
		row.Ready = true
		row.ProposedHeader = &header
	}
	return row
}

func BuildManifest(workspace string, issues []issuecontract.IssueDraft, opts Options) Manifest {
	if opts.MinScore <= 0 {
		opts.MinScore = MinScore
	}
	if opts.Review == nil {
		opts.Review = func(issue issuecontract.IssueDraft) issuecontract.Review {
			return issuecontract.ReviewIssueDraft(issue, issuecontract.Options{})
		}
	}
	if opts.Route == nil {
		opts.Route = DefaultRoute
	}
	if opts.Template == nil {
		opts.Template = issuecontract.BuildTemplateRepairPlan
	}
	limit := opts.Limit
	if limit <= 0 {
		limit = len(issues)
	}
	var lanePtr *string
	if strings.TrimSpace(opts.Lane) != "" {
		lane := strings.TrimSpace(opts.Lane)
		lanePtr = &lane
	}
	var rows []RepairRow
	examined := 0
	for _, issue := range issues {
		review := opts.Review(issue)
		route := opts.Route(issue, review)
		if lanePtr != nil && review.Lane != *lanePtr && route.Lane != *lanePtr && route.BlockedLane != *lanePtr {
			continue
		}
		if examined >= limit {
			break
		}
		examined++
		template, hasTemplate := opts.Template(issue)
		row := BuildRepairRow(issue, review, route, template, hasTemplate, opts.MinScore)
		if row != nil {
			rows = append(rows, *row)
		}
	}
	byKind := map[string]int{}
	ready := 0
	for _, row := range rows {
		byKind[row.Kind]++
		if row.Ready {
			ready++
		}
	}
	return Manifest{
		Schema:    Schema,
		AsOf:      opts.AsOf,
		Workspace: workspace,
		Lane:      lanePtr,
		Limit:     limit,
		Counts: Counts{
			CandidatesExamined: examined,
			NeedsRepair:        len(rows),
			Ready:              ready,
			ByKind:             byKind,
		},
		Issues: rows,
	}
}

func BuildActions(manifest Manifest) []Action {
	out := make([]Action, 0, len(manifest.Issues))
	for _, row := range manifest.Issues {
		reason := strings.Join(row.Reasons, ", ")
		if reason == "" {
			reason = "issue contract below spawn floor"
		}
		out = append(out, Action{
			Number: row.Number, Kind: row.Kind, Ready: row.Ready,
			Reason: reason, NextAction: row.NextAction, Cmd: nil,
		})
	}
	return out
}

func RenderMarkdown(manifest Manifest) string {
	lane := "(all lanes)"
	if manifest.Lane != nil {
		lane = *manifest.Lane
	}
	var b strings.Builder
	fmt.Fprintf(&b, "# Issue-contract repairs - %s\n\n", manifest.AsOf)
	fmt.Fprintf(&b, "**Lane:** %s  ·  **examined:** %d  ·  **needs repair:** %d  ·  **template-ready:** %d\n\n",
		lane, manifest.Counts.CandidatesExamined, manifest.Counts.NeedsRepair, manifest.Counts.Ready)
	b.WriteString("> Read-only pass. Never edits, labels, comments on, or closes an issue. `template`-kind rows carry a dry-run-computed header fix; every other kind lists the missing fields as questions for a human/agent to answer -- content is never invented here.\n\n")
	b.WriteString("## Counts by kind\n\n| kind | count |\n|---|---:|\n")
	kinds := make([]string, 0, len(manifest.Counts.ByKind))
	for kind := range manifest.Counts.ByKind {
		kinds = append(kinds, kind)
	}
	sort.Slice(kinds, func(i, j int) bool { return rank(kinds[i]) < rank(kinds[j]) })
	for _, kind := range kinds {
		fmt.Fprintf(&b, "| %s | %d |\n", kind, manifest.Counts.ByKind[kind])
	}
	b.WriteString("\n## Rows\n\n| # | kind | ready | score | title |\n|---|---|---|---:|---|\n")
	for _, row := range manifest.Issues {
		fmt.Fprintf(&b, "| #%d | %s | %t | %d | %s |\n", row.Number, row.Kind, row.Ready, row.Score, row.Title)
	}
	return b.String()
}

func FetchOpenIssues(workspace string, cap int) ([]issuecontract.IssueDraft, error) {
	if cap <= 0 {
		cap = DefaultCap
	}
	cmd := exec.Command("gh", "issue", "list", "--state", "open", "--limit", fmt.Sprint(cap), "--json", "number,title,body,labels,url")
	windowgate.ConfigureBackgroundCommand(cmd)
	cmd.Dir = workspace
	out, err := cmd.Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			return nil, fmt.Errorf("gh issue list failed (rc=%d): %s", ee.ExitCode(), strings.TrimSpace(string(ee.Stderr)))
		}
		return nil, err
	}
	var issues []issuecontract.IssueDraft
	if err := json.Unmarshal(out, &issues); err != nil {
		return nil, err
	}
	sort.Slice(issues, func(i, j int) bool { return issues[i].Number < issues[j].Number })
	return issues, nil
}

func DefaultRoute(_ issuecontract.IssueDraft, review issuecontract.Review) RouteProposal {
	if strings.TrimSpace(review.Lane) != "" {
		return RouteProposal{Lane: strings.TrimSpace(review.Lane), Confidence: "issue-contract"}
	}
	if len(review.Paths) > 0 {
		return RouteProposal{Confidence: "path-hints"}
	}
	return RouteProposal{}
}

func rank(kind string) int {
	if n, ok := kindRank[kind]; ok {
		return n
	}
	return 9
}

func contains(values []string, want string) bool {
	for _, value := range values {
		if value == want {
			return true
		}
	}
	return false
}

func scaffoldsMissingFields(kind string) bool {
	switch kind {
	case "scope", "noise", "split", "private", "other":
		return true
	default:
		return false
	}
}

func truncate(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}
